### ISSUE TITLE
Weights are adjusted each time an example has been backpropagated

### ISSUE THREAD

**Issue by KartikChugh** [KartikChugh/DeepNeuralNetworks#2]
https://github.com/KartikChugh/DeepNeuralNetworks/blob/b0508ba8d1cfea698a39373d96f3463ea6660fe0/src/kc/ml/dnn/utility/Trainer.java#L154-L172

The gradient of the cost function should be cumulatively compiled by backpropagating each example, and then used to adjust weights after all examples are done. 

Instead, the code adjusts weights after a single example is done, which is doubly incorrect, as

1. The intended implementation is gradient descent on the full dataset, not stochastic gradient descent (SGD)
2. The current implementation does not even satisfy SGD, since the gradient is compiled cumulatively (ie. each example's weight adjustments are also based on the contributions of previous examples to the gradient)
